### ISSUE TITLE
Error "Not all scopes are approved."

### ISSUE THREAD

**Issue by GEMI** [google/google-api-javascript-client#366]
I am getting error "Not all scopes are approved." when trying to refresh token. To fetch token I am doing this:
```
return Observable.create((observer) => {
    gapi.load('auth2', () => {
    
        let GoogleAuth = gapi.auth2.init({
            client_id: AUTH.GOOGLE.CLIENT_ID,
            scope: 'openid email profile'
        });

        GoogleAuth.then(() => {
            if (GoogleAuth.isSignedIn.get()) {
                let user = GoogleAuth.currentUser.get();
                this.ngZone.run(() => {
                    observer.next(user.getAuthResponse());
                });
            }
        });
    });
});
```
Then to refresh I use this code:
```
GoogleAuth.then(() => {
    if (GoogleAuth.isSignedIn.get()) {
        let user = GoogleAuth.currentUser.get();
        user.reloadAuthResponse().then((resp) => {
            this.ngZone.run(() => {
                observer.next(resp);
            });
        });
    }
});
```
And I am getting this error: {"error":"IMMEDIATE_FAILED","detail":"Not all scopes are approved."}
I hope that I am doing something incorrectly and there is a simple explanation for this.

**Comment by TMSCH** [comment]
How do you initialize the `GoogleAuth` instance in the second snippet? Is that the same set of options? By the way, `'openid email profile'` is the default scope so you don't need it.

**Comment by TMSCH** [comment]
Also, why do you reload the auth response in the second snippet? If you don't, is the error raised too?

**Comment by GEMI** [comment]
Initialization is the exact same in both cases. I refresh the token so that I could login back to my backend with the new token. Isn't this how you are supposed to get the new token if it expired?

**Comment by GEMI** [comment]
I haven't tried not to reload. How would I get a new token in that case?

**Comment by TMSCH** [comment]
When it initializes, it signs in a previously signed in user if any, so you can directly use https://developers.google.com/identity/sign-in/web/reference#googleusergetauthresponseincludeauthorizationdata

**Comment by GEMI** [comment]
Hi again, after changing refresh code from reloadAuthResponse to this
```
gapi.load('auth2', () => {

    let GoogleAuth = gapi.auth2.init({
        client_id: AUTH.GOOGLE.CLIENT_ID
    });

    GoogleAuth.then(() => {
        if (GoogleAuth.isSignedIn.get()) {
            let user = GoogleAuth.currentUser.get();
            this.ngZone.run(() => {
                observer.next(user.getAuthResponse());
            });
        }
    },
    (error) => {
        throw new Error('Error: ' + error);
    });
});
```
I am now getting an error from google:
`{"message":"Token validation error: {\n \"error_description\": \"Invalid Value\"\n}\n"}`

Example token:
```
https://www.googleapis.com/oauth2/v3/tokeninfo?id_token=eyJhbGciOiJSUzI1NiIsImtpZCI6IjI1OWY0MzMwY2UwZDI1NWFkNWM5MGFmNzJmODMwMDYwOTIwYjY2MjAifQ.eyJhenAiOiIyMzQ0OTI1OTY0NDMtcWY2dTRxcXBxaG42bXU0NmNiaWdwczM3YzFoMGRia3YuYXBwcy5nb29nbGV1c2VyY29udGVudC5jb20iLCJhdWQiOiIyMzQ0OTI1OTY0NDMtcWY2dTRxcXBxaG42bXU0NmNiaWdwczM3YzFoMGRia3YuYXBwcy5nb29nbGV1c2VyY29udGVudC5jb20iLCJzdWIiOiIxMTEyMjM1NTkyMTk2MTA4NjA4MzUiLCJlbWFpbCI6ImcubWlrZWxlbmFzQGdtYWlsLmNvbSIsImVtYWlsX3ZlcmlmaWVkIjp0cnVlLCJhdF9oYXNoIjoienVfcV9kR0Y2RHdsVkNsUFZGRkM2QSIsImlzcyI6ImFjY291bnRzLmdvb2dsZS5jb20iLCJqdGkiOiJiY2MxN2U3MTU2ZTI5ZTY4MDVlMTQ4YzdlN2VmZThmMWQxMDUxNjZmIiwiaWF0IjoxNTEyODYxNTQwLCJleHAiOjE1MTI4NjUxNDAsIm5hbWUiOiJHZWRpbWluYXMgTWlrZWxlbmFzIiwicGljdHVyZSI6Imh0dHBzOi8vbGg0Lmdvb2dsZXVzZXJjb250ZW50LmNvbS8tQmRHV0JlWFpweTQvQUFBQUFBQUFBQUkvQUFBQUFBQUFBc2svdWpNeFpmN056aFUvczk2LWMvcGhvdG8uanBnIiwiZ2l2ZW5fbmFtZSI6IkdlZGltaW5hcyIsImZhbWlseV9uYW1lIjoiTWlrZWxlbmFzIiwibG9jYWxlIjoiZW4ifQ.p1WDm5U40CDf7YizpFeZihO2SM8QJCEFpjEmBQjJkPJ1-nuXQ2o9BA7ZGBS951Xaftp1JG4TLh8EZEBiIeN-8ArSSpf9bgjCJiHwV78gAiqhDPzRQmfF3X4qCUEZEH0cfuh5GoBKPtO6b-yYVmwf4w92YuHQXlLsf-w9SAchY-FzGhTkNd2L8FOjwa694R_Go2YhQw93WIs2EqylhSJ594VaMc9Te5jaw9NISXE45GhKvjTLUmyMB0Ng2AbMk-W3hJ1bY2xW9n_3FbVfz1KIFKNN3LrHun5yReR-sQgwCj9oJzRCDYxsqhVgF7Sei4UrsY41ELkXcgfIKEKpAb9bMA
```

**Comment by TMSCH** [comment]
Could you post the snippet of code that you use to call the `tokeninfo` endpoint?

**Comment by GEMI** [comment]
Well we do it back-end. What could this be potentially?
```
const string url = "https://www.googleapis.com/oauth2/v3/tokeninfo?id_token=";
var result = await _client.GetAsync(url + token);
var content = await result.Content.ReadAsStringAsync();
var tokenInfo = JsonConvert.DeserializeObject<TokenInfo>(content);
```

**Comment by TMSCH** [comment]
`getAuthResponse` returns an object with `id_token` and `access_token` as properties. Are you properly extracting the ID Token?

**Comment by GEMI** [comment]
Well there is not much to extract just `response.id_token;` Regarding the token as you can see from the previous example it seems that the token is being sent, at least it looks like a token. Also I want to emphasize that this happens only when token is expired, because the first time everything works fine. Also when I used the `reloadAuthResponse` call this error never happened.
Maybe you could point me to some resource or repository that correctly uses GAPI to login and re-login on token expiration?
Also if you have something in mind regarding what could help to debug this issue please let me know, I could try some more stuff out. Thanks.

**Comment by TMSCH** [comment]
@GEMI it's expected that this call fails when the `id_token` has expired. Are you saying that it's when you re-login more than 1 hour after the first time that this error happens? Unless you call `reloadAuthResponse`?

**Comment by GEMI** [comment]
Yes, exactly. So am I supposed to use `reloadAuthResponse` to refresh token or not? Because I thought you said it is not needed. Sorry I am confused with the workflow here.

**Comment by TMSCH** [comment]
You should not have to, @GEMI, as the library should not reuse a cached expired token. It seems there might be a bug here. Is your computer clock properly set and synced?

After you have signed in, please check the Application tab of the Developers Console (in Chrome). There is a SessionStorage section. Click the `https://accounts.google.com` sub-section. There will be an entry with a key starting with `oauth2_tr`. The value saved is the full token response. It has an `expires_at` field. Could you check that the following JS code returns something in the order of ~3600 (right after signing in for the first time):

```js
(expires_at - new Date().getTime()) / 1000
```

To confirm: the issue happens only in the same tab or not? If you open a new tab, the user should be automatically signed in and a new and valid token should be issued. Is that the case? (if you open a new tab you actually don't have to wait for an hour, the cached token is per tab)

**Comment by GEMI** [comment]
Hi, yes the time after login is approx 3600. Yes the issue happens when trying to reload in the same tab. However, I will double check to see if this happens with a new tab as well.

**Comment by dvanderb** [comment]
This looks like a question about Google Login and Authentication. Please address this question to the google-signin tag on stackoverflow(https://stackoverflow.com/questions/tagged/google-signin)

If you believe this is in error, please reopen the issue.
